polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp

```diff
@@ -0,0 +1,168 @@
+}
+
+} // namespace polygeist
+} // namespace mlir
+
+void UniformityLattice::onUpdate(DataFlowSolver *solver) const {
+  Lattice::onUpdate(solver);
+
+  auto value = point.get<Value>();
+  Operation *op = value.getDefiningOp();
+
+  if (op) {
+    if (auto branch = dyn_cast<RegionBranchOpInterface>(op)) {
+      for (Value operand : branch->getOperands()) {
+        auto *uniformity = solver->lookupState<UniformityLattice>(operand);
+        assert(uniformity && "Expecting a valid uniformity");
+
+        auto *lattice = solver->getOrCreateState<UniformityLattice>(value);
+        return solver->propagateIfChanged(lattice, lattice->join(*uniformity));
+      }
+    }
+  }
+
+  // Note: required to get the dataflow framework to visit nested regions.
+  auto *cv = solver->getOrCreateState<Lattice<ConstantValue>>(value);
+  return solver->propagateIfChanged(
+      cv, cv->join(ConstantValue::getUnknownConstant()));
+}
+
+void Uniformity::print(raw_ostream &os) const { os << *this; }
+
+//===----------------------------------------------------------------------===//
+// UniformityAnalysis
+//===----------------------------------------------------------------------===//
+
+void UniformityAnalysis::visitOperation(
+    Operation *op, ArrayRef<const UniformityLattice *> operands,
+    ArrayRef<UniformityLattice *> results) {
+  LLVM_DEBUG(llvm::dbgs() << "UA: Visiting operation: " << *op << "\n");
+
+  // If the lattice on any operand is uninitialized, bail out.
+  if (llvm::any_of(operands, [](const UniformityLattice *lattice) {
+        return lattice->getValue().isUninitialized();
+      })) {
+    return;
+  }
+
+  // These operations always yield non-uniform result(s).
+  if (isa<sycl::SYCLItemGetIDOp, sycl::SYCLNDItemGetGlobalIDOp>(op)) {
+    for (UniformityLattice *result : results)
+      propagateIfChanged(result, result->join(Uniformity::getNonUniform()));
+    return;
+  }
+
+  auto anyOperandUniformityIs = [&](Uniformity::Kind kind) {
+    return llvm::any_of(operands, [&](const UniformityLattice *lattice) {
+      return lattice->getValue().getKind() == kind;
+    });
+  };
+
+  // If any operand uniformity is unknown/non-uniform the result(s) are also
+  // unknown/non-uniform.
+  if (anyOperandUniformityIs(Uniformity::Kind::Unknown)) {
+    setAllToEntryStates(results);
+    return;
+  }
+  if (anyOperandUniformityIs(Uniformity::Kind::NonUniform)) {
+    for (UniformityLattice *result : results)
+      propagateIfChanged(result, result->join(Uniformity::getNonUniform()));
+    return;
+  }
+
+  assert(llvm::all_of(operands,
+                      [&](const UniformityLattice *lattice) {
+                        return lattice->getValue().getKind() ==
+                               Uniformity::Kind::Uniform;
+                      }) &&
+         "Expecting all operands to be uniform");
+
+  // A memory side effects free operation that has uniform operands yields
+  // uniform result(s).
+  if (isMemoryEffectFree(op)) {
```

Comment:
This is incorrect, as there are memory effect free operations with non-uniform results, e.g., [`gpu.thread_id`](https://mlir.llvm.org/docs/Dialects/GPU/#gputhread_id-gputhreadidop).

Reply:
Right, the idea is to catch those operations at line 92 (like we do for the SYCL operations that yield a thread id). I didn't address other possible sources of non-uniformity in this PR (is part 1:-)